File: ggcommons/config/manager/shadow_config_manager.py
import json
import logging
from abc import ABC

from awsiot.greengrasscoreipc.clientv2 import GreengrassCoreIPCClientV2
from awsiot.greengrasscoreipc.model import ReceiveMode, SubscriptionResponseMessage
from ggcommons.config.manager.config_manager import ConfigManager

logger = logging.getLogger("ShadowConfigManager")


class ShadowConfigManager(ConfigManager, ABC):
    _SHADOW_TOPIC_TEMPLATE = "$aws/things/{}/shadow/name/{}/"
    _ALL_SHADOW_TOPIC_TEMPLATE = "$aws/things/{}/shadow/name/{}/+/+"
    _DEFAULT_CONFIGURATION = {
        "logging": {},
        "source": {},
        "heartbeat": {},
        "component": {"global": {}, "instances": []},
    }

    def __init__(self, component_name: str, shadow_name: str):
        super().__init__(component_name)
        self._shadow_name = shadow_name if shadow_name is not None else component_name
        self._ipc_client = GreengrassCoreIPCClientV2()
        self._shadow_topic_prefix = ShadowConfigManager._SHADOW_TOPIC_TEMPLATE.format(
            self.get_thing_name(), shadow_name
        )
        self._subscribe_to_shadow_topics()
        self.init()

    def _subscribe_to_shadow_topics(self):
        logger.debug("Subscribing to shadow topics")
        try:
            shadow_update_delta_topic = (
                ShadowConfigManager._ALL_SHADOW_TOPIC_TEMPLATE.format(
                    self.get_thing_name(), self._shadow_name
                )
            )
            _, operation = self._ipc_client.subscribe_to_topic(
                topic=shadow_update_delta_topic,
                receive_mode=ReceiveMode.RECEIVE_MESSAGES_FROM_OTHERS,
                on_stream_closed=None,
                on_stream_error=None,
                on_stream_event=self._on_shadow_event,
            )
        except Exception as e:
            logger.error(f"Failed to subscribe to shadow topics: {e}")

    def _on_shadow_event(self, event: SubscriptionResponseMessage) -> None:
        payload_str = str(event.binary_message.message, "utf-8")
        topic_parts = event.binary_message.context.topic.split("/")
        action = topic_parts[len(topic_parts) - 2]
        result = topic_parts[len(topic_parts) - 1]
        logger.debug(
            f"Received shadow message for shadow action '{action}' result '{result}'. Payload: {payload_str}"
        )

        if action == "get" and result == "rejected":
            logger.warning(
                f"Named shadow document {self._shadow_name} does not exist. Creating default configuration."
            )
            self._report_updated_configuration(
                ShadowConfigManager._DEFAULT_CONFIGURATION
            )
        elif action == "update" and result == "delta":
            payload_json = json.loads(payload_str)
            desired_doc = payload_json["state"]
            if desired_doc is not None:
                logger.debug(f"Desired document: {desired_doc}")
                component_config = json.loads(desired_doc["ComponentConfig"])
                self.configuration_changed(component_config)
                self._report_updated_configuration(component_config)
        # else:
        #     logger.info(f"Received message for shadow action '{action}' result '{result}'.")

    def _load_configuration(self) -> dict:
        logger.debug(f"Loading configuration from named shadow ('{self._shadow_name}')")
        config = self._get_configuration()
        if config is not None:
            self._report_updated_configuration(config)
        return config

    def _report_updated_configuration(self, config: dict) -> None:
        shadow_doc = {
            "state": {
                "reported": {
                    "ComponentConfig": json.dumps(
                        config, indent=None, separators=(",", ":")
                    )
                }
            }
        }
        logger.debug(
            f"Reporting updated configuration to named shadow document '{self._shadow_name}': {shadow_doc}"
        )
        try:
            self._ipc_client.update_thing_shadow(
                thing_name=self.get_thing_name(),
                shadow_name=self._shadow_name,
                payload=json.dumps(shadow_doc).encode("utf-8"),
            )
        except Exception as e:
            logger.error(f"Failed to report updated configuration: {e}")

    def _get_configuration(self):
        try:
            response = self._ipc_client.get_thing_shadow(
                thing_name=self.get_thing_name(), shadow_name=self._shadow_name
            )
            if response.payload is not None and len(response.payload) > 0:
                payload = str(response.payload, "utf-8")
                payload_json = json.loads(payload)
                state_doc = payload_json["state"]
                if "desired" in state_doc:
                    return json.loads(state_doc["desired"]["ComponentConfig"])
                elif "reported" in state_doc:
                    return json.loads(state_doc["reported"]["ComponentConfig"])
            else:
                logger.warning(
                    f"Named shadow document '{self._shadow_name}' does not exist or is empty"
                )
            return None
        except Exception as e:
            logger.error(f"_get_configuration: Failed to get configuration: {str(e)}")
            return None

    def get_config_source(self) -> str:
        return f"Named Shadow (shadow name: {self._shadow_name})"

File: src/main/java/com/aws/proseve/ggcommons/config/manager/ShadowConfigManager.java
package com.aws.proseve.ggcommons.config.manager;

import com.aws.proserve.ggcommons.utils.Utils;
import com.github.cliftonlabs.json_simple.JsonObject;
import com.github.cliftonlabs.json_simple.Jsoner;
import software.amazon.awssdk.aws.greengrass.GreengrassCoreIPCClientV2;
import software.amazon.awssdk.aws.greengrass.SubscribeToTopicResponseHandler;
import software.amazon.awssdk.aws.greengrass.model.*;
import software.amazon.awssdk.eventstreamrpc.StreamResponseHandler;

import java.nio.charset.StandardCharsets;

class ShadowConfigManager extends ConfigManager implements StreamResponseHandler<SubscriptionResponseMessage>
{
    protected static final String SHADOW_TOPIC_TEMPLATE = "$aws/things/%s/shadow/name/%s/";
    protected static final String ALL_SHADOW_TOPIC_TEMPLATE = "$aws/things/%s/shadow/name/%s/+/+";
    protected final String shadowTopicPrefix;
    private final String shadowName;
    GreengrassCoreIPCClientV2 ipcClient;

    ShadowConfigManager(String componentName, String shadowName)
    {
        super(componentName);
        this.shadowName = shadowName;
        this.shadowTopicPrefix = String.format(SHADOW_TOPIC_TEMPLATE, getThingName(), shadowName);
        connectToIPC();
        subscribeShadowTopics();
        init();
    }

    @Override
    protected JsonObject loadConfiguration()
    {
        JsonObject retVal = null;
        LOGGER.debug("Loading configuration from named shadow ('{}')", shadowName);
        String componentConfigStr = getConfiguration();
        if (componentConfigStr != null)
        {
            LOGGER.debug("Configuration loaded from named shadow ('{}')", shadowName);
            retVal = Utils.destringify(componentConfigStr);
            reportUpdatedConfiguration(componentConfigStr);
        }
        return retVal;
    }

    @Override
    protected String getConfigSource()
    {
        return String.format("Named shadow (shadow name: '%s')", shadowName);
    }

    private void reportUpdatedConfiguration(String componentConfig)
    {
        JsonObject shadowDoc = new JsonObject();
        JsonObject stateDoc = new JsonObject();
        JsonObject reportedDoc = new JsonObject();

        LOGGER.trace("Updating com.aws.proseve.ggcommons.config to:{}", componentConfig);
        reportedDoc.put("ComponentConfig", componentConfig);
        stateDoc.put("reported", reportedDoc);
        shadowDoc.put("state", stateDoc);

        try
        {
            UpdateThingShadowRequest updateRequest = new UpdateThingShadowRequest().withThingName(getThingName())
                                                                                   .withShadowName(shadowName)
                                                                                   .withPayload(shadowDoc.toJson().getBytes(StandardCharsets.UTF_8));
            UpdateThingShadowResponse updateResponse = ipcClient.updateThingShadow(updateRequest);
            LOGGER.trace("Update shadow response: {}", updateResponse.toString());
        }
        catch (InterruptedException e)
        {
            LOGGER.error("Shadow update failed: {}", e.getMessage());
        }
    }

    private void connectToIPC()
    {
        try
        {
            ipcClient = GreengrassCoreIPCClientV2.builder().build();
        }
        catch (Exception e)
        {
            LOGGER.fatal("Unable to connect to Greengrass IPC.");
            System.exit(5);
        }
    }

    private String getConfiguration()
    {
        String retVal = null;
        try
        {
            GetThingShadowRequest request = new GetThingShadowRequest().withThingName(getThingName())
                                                                       .withShadowName(shadowName);
            GetThingShadowResponse response = ipcClient.getThingShadow(request);
            LOGGER.trace("Get shadow response: {}", response.toString());
            byte[] payload = response.getPayload();
            if (payload != null && payload.length > 0)
            {
                String payloadAsString = new String(payload, StandardCharsets.UTF_8);
                JsonObject shadowDoc = (JsonObject) Jsoner.deserialize(payloadAsString);
                JsonObject desiredDoc = (JsonObject) ((JsonObject) shadowDoc.get("state")).get("desired");
                retVal = desiredDoc.get("ComponentConfig").toString();
            }
            else
            {
                LOGGER.info("Named shadow document {} is empty", shadowName);
            }
        }
        catch (Exception e)
        {
            LOGGER.error("Failed to get named shadow document {}: {}", shadowName, e.getMessage());
        }
        return retVal;
    }

    private void subscribeShadowTopics()
    {
        String shadowUpdateDeltaTopic = String.format(ALL_SHADOW_TOPIC_TEMPLATE, getThingName(), shadowName);
        try
        {
            SubscribeToTopicRequest subRequest = new SubscribeToTopicRequest().withTopic(shadowUpdateDeltaTopic)
                                                                              .withReceiveMode(ReceiveMode.RECEIVE_MESSAGES_FROM_OTHERS);
            GreengrassCoreIPCClientV2.StreamingResponse<SubscribeToTopicResponse,
                    SubscribeToTopicResponseHandler> response =
                    ipcClient.subscribeToTopic(subRequest, this);
            LOGGER.info("Subscribed to IPC messages for shadow updates.");
        }
        catch (Exception e)
        {
            LOGGER.error("Failed to subscribe to IPC messages for shadow updates: {}", e.getMessage());
        }
    }

    @Override
    public void onStreamEvent(SubscriptionResponseMessage subscriptionResponseMessage)
    {
        try {
            BinaryMessage binaryMessage = subscriptionResponseMessage.getBinaryMessage();
            String message = new String(binaryMessage.getMessage(), StandardCharsets.UTF_8);
            String topic = binaryMessage.getContext().getTopic();
            String[] topicParts = topic.split("/");
            String action = topicParts[topicParts.length - 2];
            String result = topicParts[topicParts.length - 1];

            if (action.equals("get") && result.equals("rejected")) {
                LOGGER.warn("Named shadow document {} does not exist.  Creating default configuration.", shadowName);
                reportUpdatedConfiguration(getDefaultConfig().toJson());
            } else if (action.equals("update") && result.equals("accepted")) {
                LOGGER.debug("Received update/accepted message.  Attempting to apply changes. message:  {}", message);
                String decodedBinaryPayload = new String(subscriptionResponseMessage.getBinaryMessage().getMessage(), StandardCharsets.UTF_8);
                JsonObject payload = (JsonObject) Jsoner.deserialize(decodedBinaryPayload);
                JsonObject desiredDoc = (JsonObject) ((JsonObject) (payload.get("state"))).get("desired");
                if (desiredDoc != null)
                {
                    String componentConfigStr = desiredDoc.get("ComponentConfig").toString();
                    JsonObject componentConfig = Utils.destringify(componentConfigStr);
                    configurationChanged(componentConfig);
                    reportUpdatedConfiguration(componentConfigStr);
                }
            } else if (action.equals("update") && result.equals("delta")) {
                LOGGER.warn("Received update/delta message. {}", message);
            }
            else
            {
                LOGGER.debug("Received new shadow message on topic {}. Ignoring", topic);
            }
        } catch (Exception e) {
            LOGGER.error("Exception occurred while processing subscription response: {}, {}", e.getMessage(), e.getStackTrace());
        }
    }

    @Override
    public boolean onStreamError(Throwable throwable)
    {
        LOGGER.error("Error on IPC stream for subscription to shadow updates: {}", throwable.getMessage());
        return false;
    }

    @Override
    public void onStreamClosed()
    {
        LOGGER.info("IPC stream for subscription to shadow updates closed (unsubscribed)");
    }
}
